File: creational/abstract_factory/llm_providers/typescript/llm_providers.test.ts
import {
  LLMProviderFactory,
  OpenAIFactory,
  AnthropicFactory,
  GeminiFactory,
  OllamaFactory,
} from "./llm_providers";

// Helper function to get concrete class names (less brittle than direct import)
// This relies on the structure within llm_providers.ts
// It's a bit of a workaround because TS interfaces don't exist at runtime.
// A better approach in a larger project might be distinct files per class.
const getImplementationName = (obj: any): string =>
  obj?.constructor?.name ?? "unknown";

describe("LLM Provider Abstract Factories", () => {
  // Helper function to test a factory
  async function testFactory(
    factory: LLMProviderFactory,
    expectedClientClass: string,
    expectedConfigClass: string,
    expectedModel: string,
    expectedCompletionSubstring: string
  ) {
    const client = factory.createClient();
    const config = factory.createConfiguration(); // Create separately for assertions

    // Check types using constructor names as a proxy
    // expect(client).toBeInstanceOf(expectedClientClass); // This won't work directly with class vars
    expect(getImplementationName(client)).toBe(expectedClientClass);
    expect(getImplementationName(config)).toBe(expectedConfigClass);

    // Check config details
    expect(config.getModel()).toBe(expectedModel);

    // Check if client holds the correct config internally (if exposed)
    // Assuming the client has a public `config` property for this test
    const clientInternalConfig = (client as any).config;
    expect(clientInternalConfig).toBeDefined();
    expect(getImplementationName(clientInternalConfig)).toBe(
      expectedConfigClass
    );
    expect(clientInternalConfig.getModel()).toBe(expectedModel);

    // Test dummy generation (async)
    const completion = await client.generate(
      `test prompt ${expectedClientClass}`
    );
    expect(completion).toContain(expectedCompletionSubstring);
  }

  test("OpenAIFactory should create OpenAI products", async () => {
    await testFactory(
      new OpenAIFactory(),
      "OpenAIClient", // Expected constructor name
      "OpenAIConfiguration",
      "gpt-4o",
      "OpenAI completion"
    );
  });

  test("AnthropicFactory should create Anthropic products", async () => {
    await testFactory(
      new AnthropicFactory(),
      "AnthropicClient",
      "AnthropicConfiguration",
      "claude-3.7-sonnet",
      "Anthropic completion"
    );
  });

  test("GeminiFactory should create Gemini products", async () => {
    await testFactory(
      new GeminiFactory(),
      "GeminiClient",
      "GeminiConfiguration",
      "gemini-2.5-pro",
      "Gemini completion"
    );
  });

  test("OllamaFactory should create Ollama products", async () => {
    await testFactory(
      new OllamaFactory(),
      "OllamaClient",
      "OllamaConfiguration",
      "llama4",
      "Ollama completion"
    );
  });
});

File: creational/abstract_factory/llm_providers/go/llm_providers/llm_providers_test.go
package llm_providers

import (
	"reflect" // Use reflection for type checking
	"strings"
	"testing"
)

// Helper function to test factories using reflection for type checks
func testFactory(t *testing.T, factory LLMProviderFactory, expectedClientType reflect.Type, expectedConfigType reflect.Type, expectedModel string, expectedCompletionSubstring string) {
	// Suppress verbose output during tests if needed, but keep for demonstration
	// t.Logf("Testing factory: %T", factory)

	client := factory.CreateClient()
	config := factory.CreateConfiguration() // Create separately for checks

	// Check client type using reflection
	clientType := reflect.TypeOf(client)
	if clientType != expectedClientType {
		t.Errorf("Factory %T created client of type %v, expected %v", factory, clientType, expectedClientType)
	}

	// Check config type using reflection
	configType := reflect.TypeOf(config)
	if configType != expectedConfigType {
		t.Errorf("Factory %T created config of type %v, expected %v", factory, configType, expectedConfigType)
	}

	// Check config model
	if config.GetModel() != expectedModel {
		t.Errorf("Factory %T created config with model '%s', expected '%s'", factory, config.GetModel(), expectedModel)
	}

	// Check client uses the correct config implicitly
	// We need to access the private 'Config' field via reflection for a robust test
	clientValue := reflect.ValueOf(client)
	// If client is a pointer, get the element it points to
	if clientValue.Kind() == reflect.Ptr {
		clientValue = clientValue.Elem()
	}
	// Access the 'Config' field (now exported)
	internalConfigField := clientValue.FieldByName("Config")
	if !internalConfigField.IsValid() {
		t.Fatalf("Factory %T client type %T does not have an accessible 'Config' field for testing", factory, client)
	}
	// This should no longer panic
	internalConfig := internalConfigField.Interface().(LLMConfiguration) // Assert to interface

	internalConfigType := reflect.TypeOf(internalConfig)
	if internalConfigType != expectedConfigType {
		t.Errorf("Factory %T client internal config is of type %v, expected %v", factory, internalConfigType, expectedConfigType)
	}
	if internalConfig.GetModel() != expectedModel {
		t.Errorf("Factory %T client internal config has model '%s', expected '%s'", factory, internalConfig.GetModel(), expectedModel)
	}


	// Test dummy generation
	completion, err := client.Generate("test prompt")
	if err != nil {
		t.Errorf("Factory %T client generate failed: %v", factory, err)
	}
	if !strings.Contains(completion, expectedCompletionSubstring) {
		t.Errorf("Factory %T client generated '%s', expected substring '%s'", factory, completion, expectedCompletionSubstring)
	}
}

// --- Test Cases ---

func TestOpenAIFactory(t *testing.T) {
	factory := &OpenAIFactory{}
	// Get expected types using reflection on pointers to the structs
	expectedClientType := reflect.TypeOf(&openAIClient{})
	expectedConfigType := reflect.TypeOf(&openAIConfiguration{})
	testFactory(t, factory, expectedClientType, expectedConfigType, "gpt-4o", "OpenAI completion")
}

func TestAnthropicFactory(t *testing.T) {
	factory := &AnthropicFactory{}
	expectedClientType := reflect.TypeOf(&anthropicClient{})
	expectedConfigType := reflect.TypeOf(&anthropicConfiguration{})
	testFactory(t, factory, expectedClientType, expectedConfigType, "claude-3.7-sonnet", "Anthropic completion")
}

func TestGeminiFactory(t *testing.T) {
	factory := &GeminiFactory{}
	expectedClientType := reflect.TypeOf(&geminiClient{})
	expectedConfigType := reflect.TypeOf(&geminiConfiguration{})
	testFactory(t, factory, expectedClientType, expectedConfigType, "gemini-2.5-pro", "Gemini completion")
}

func TestOllamaFactory(t *testing.T) {
	factory := &OllamaFactory{}
	expectedClientType := reflect.TypeOf(&ollamaClient{})
	expectedConfigType := reflect.TypeOf(&ollamaConfiguration{})
	testFactory(t, factory, expectedClientType, expectedConfigType, "llama4", "Ollama completion")
}
